api/src/main/java/org/apache/iceberg/VariantLike.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.math.BigDecimal;
+
+/**
+ * Interface for accessing Variant fields.
+ *
+ * <p>This interface supports accessing data in top-level fields, not in nested fields.
```

Comment:
I don't think we need this comment since I believe it should be allowed?

Reply:
Actually what I'm thinking is: we only access the top-level fields to get Variant, and then recursively you get subfield from that Variant.

So for a path `$.a.b`, we get Variant for `$.a` and then from that Variant get `$.a.b` Variant.

Reply:
My confusion with the recursive interface is that we actually don't have variant as a subfield, we have a well defined type in all cases as defined by the metadata. My mental model is that a variant is essentially
{path:type:offset,path:type:offset, .....} {value blob}

So I thought the most natural interface would be one that mimics that pattern

Reply:
The variant binary encoding for objects is more based on fields, not entire paths.
```
{field1, field2, field3, offset1, offset2, offset3, value1, value2, value3}
```
If the `valueN` is an object, the encoding scheme would continue as above.

A sub-value of a variant, is also a valid variant value, so `valueN` is valid Variant binary.